### PR TITLE
Fixed a few issues with exception handling in the resource cache

### PR DESCRIPTION
**Pull request by kiiadi**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Description

Previously exceptions were getting cached, and bubbled as unhandled exceptions on subsequent calls. This fixes the following:
- exceptions are no longer considered valid "current values" when evaluating whether to return a cached value
- exceptions are pruned from the cache on maintenance runs

## Related Issue(s)

#3098 

## Testing

Added new tests to cover missing scenarios.

## Checklist

- [ ] I have read the **README** document
- [ ] I have read the **CONTRIBUTING** document
- [ ] Local run of `gradlew check` succeeds
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if change is customer facing in the IDE.
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt [hunk 1]
```diff
@@ -340,21 +341,27 @@ class DefaultAwsResourceCache(
...
-                    exceededWeight -= value.weight
-                }
-            }
+            doRunCacheMaintenance()
         } finally {
             scheduleCacheMaintenance()
         }
     }
 
+    @VisibleForTesting
```
Comment (kiiadi) [inline jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt:350]: I do hate this - but for internal state like this I just can't justify injection ... I'm very conflicted about it though

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt [hunk 2]
```diff
@@ -475,6 +480,50 @@ class AwsResourceCacheTest {
...
+        val latch = CountDownLatch(1)
+        whenever(mockResource.fetch(any(), any())).then {
+            latch.await(500, TimeUnit.MILLISECONDS)
+            throw RuntimeException("Boom")
+        }
+
+        val first = sut.getResource(mockResource, connectionSettings)
+        val second = sut.getResource(mockResource, connectionSettings)
+        latch.countDown()
+        assertThat(first).wait().hasException.withFailMessage { "" }
```
Comment (kiiadi) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt:494]: oops - gotta fix these messges

Comment (rli) [inline jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt:494]: assert that it's the same exception and not 2 different calls?

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt (hunk 1 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt:350]: i think we can lint for this if that makes you feel better

**Review by rli — APPROVED** [review]

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt (hunk 1 above)
Comment (kiiadi) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt:350]: lint for what? that it's only used in tests?

File: jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt (hunk 2 above)
Comment (kiiadi) [reply jetbrains-core/tst/software/aws/toolkits/jetbrains/core/AwsResourceCacheTest.kt]: added a verify that the mock only gets hit once

File: jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt (hunk 1 above)
Comment (rli) [reply jetbrains-core/src/software/aws/toolkits/jetbrains/core/AwsResourceCache.kt:350]: yup, we have a number of `@VisibleForTesting` peppered around the code base but i'm not actually sure those are being respected

**Pull request merged**